src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs

```diff
@@ -0,0 +1,309 @@
+        {
+            OperationName = operationName;
+        }
+
+        /// <summary>
+        /// Update the Activity to have a tag with an additional 'key' and value 'value'.
+        /// This shows up in the 'Tags' eumeration.   It is meant for information that
+        /// is useful to log but not needed for runtime control (for the latter, use Baggage)
+        /// </summary>
+        /// <returns>'this' for convinient chaining</returns>
+        public Activity AddTag(string key, string value)
+        {
+            _tags = new KeyValueListNode() { keyValue = new KeyValuePair<string, string>(key, value), Next = _tags };
+            return this;
+        }
+
+        /// <summary>
+        /// Update the Activity to have baggage with an additional 'key' and value 'value'.
+        /// This shows up in the 'Baggage' eumeration as well as the 'GetBaggageItem' API.
+        /// Baggage is mean for information that is needed for runtime control.   For information 
+        /// that is simply useful to show up in the log with the activity use Tags.   
+        /// Returns 'this' for convinient chaining.
+        /// </summary>
+        /// <returns>'this' for convinient chaining</returns>
+        public Activity AddBaggage(string key, string value)
+        {
+            _baggage = new KeyValueListNode() { keyValue = new KeyValuePair<string, string>(key, value), Next = _baggage };
+            return this;
+        }
+
+        /// <summary>
+        /// Updates the Activity To indicate that the activity with ID 'parentID' 
+        /// caused this activity.   This is only intended to be used at 'boundary' 
+        /// scenarios where an activity from another process loggically started 
+        /// this activity. The Parent ID shows up the Tags (as well as the ParentID 
+        /// property), and can be used to reconstruct the causal tree.  
+        /// Returns 'this' for convinient chaining.
+        /// </summary>
+        public Activity SetParentId(string parentId)
+        {
+            ParentId = parentId;
+            return this;
+        }
+
+        /// <summary>
+        /// Update the Activity to set start time
+        /// </summary>
+        /// <param name="startTimeUtc">Activity start time in UTC (Greenwitch Mean Time)</param>
+        /// <returns>'this' for convinient chaining</returns>
+        public Activity SetStartTime(DateTime startTimeUtc)
+        {
+            StartTimeUtc = startTimeUtc;
+            return this;
+        }
+        /// <summary>
+        /// Update the Activity to set <see cref="Duration"/>
+        /// as a difference between <see cref="StartTimeUtc"/>
+        /// and given stop timestamp
+        /// </summary>
+        /// <param name="endTimeUtc">Activity stop time in UTC (Greenwitch Mean Time)</param>
+        /// <returns>'this' for convinient chaining</returns>
+        public Activity SetEndTime(DateTime endTimeUtc)
+        {
+            Duration = endTimeUtc - StartTimeUtc;
```

Comment:
Stop will be called **after** DiagnosticSource.Write will notify user about activity being stopping. 

Here I'd like to give control to user: if he wants to set stop time in the past, I don't see reasons to prevent it. I agree though we should insist on Utc Kind.